### DISCUSSION TITLE
save and restore columns in <p:dataDable>

### DISCUSSION THREAD

**Discussion by bgahn-materna** [primefaces/community#1029] [PrimeFaces] [answered]
Hi all,

we are building a frontend based on filters and datatables for corresponding results. Since most pages look more or less alike each other, we decided to use a dynamic column model and defined the datatable like this:
`<p:dataTable ...> <p:columns value="#{pageBean.columns}" ...>CONTENT_OF_COLUMNS</p:columns></p:dataTable>`

Users are supposed to customize their view of the table in a setting and then save it for later use in another session. Individual settings include pagination, sorting, column visibility, column width and column order. We try to achieve this through <p:dataTable> and its features.

What we have thus far is customization of column width and visibility, as well as sorting and pagination. This is done by accessing the datatable like such:

`final DataTable dataTable = (DataTable) FacesContext.getCurrentInstance().getViewRoot()findComponent(dataTableId);`
`final DataTableSettings<E> dataTableSettings = new DataTableSettings<>(dataTable.getRows(), dataTable.getColumnMeta(), dataTable.getSortByAsMap());`

We then continue to serialize this information into JSON (utilizing Jackson's `ObjectMapper`) and save the DataTableSetting-object to our database, and upon reloading the page we deserialize the JSON back into the settings, then reapply the values to the datatable:

`final DataTable dataTable = (DataTable) FacesContext.getCurrentInstance()getViewRoot().findComponent(dataTableId);`
`final DataTableSettings<E> dataTableSettings = mapKonfigToDataTableSettings(jsonString);`
`dataTable.setRows(dataTableSettings.getRows());`
`dataTable.setColumnMeta(dataTableSettings.getColumnMetaMap());`
`dataTable.setSortByAsMap(dataTableSettings.getSortMetaMap());`

Sidenote: We try to avoid component binding because our backing beans require to be `Serializable` because of `@ViewScoped` lifespan.
This is working great by itself. However, once we start enabling reordering of columns via `draggableColumns`, problems arise. In concept it should be enough to include the underlying column model from the backing bean in the serialization and later restore it to the bean itself. To make that work, we added an AJAX-listener to the columnReorderEvent:
`<p:ajax event="colReorder" listener="#{pageController.onColumnReorder}"/>`
This was mentioned [here](https://forum.primefaces.org/viewtopic.php?t=20718) and [here](https://forum.primefaces.org/viewtopic.php?t=29571). We verified this and can successfully handle this event to update the column model. However, if we restore the column order upon loading the configuration, things start to break apart if the order does not match the initial order:

* column header is not correctly displayed and does not match the data that is rendered underneath it
* information about columns being sortable or not seem to get lost (columns that should not be sortable become sortable)
* sort criteria associated with the column don't seem to match anymore

We also tried to serialize the `Columns`-object of the datatable, but this cannot be correctly done because one has to exclude too many references in order to avoid infinite chains and (failing) generic asserts on getters of `UIComponent`s.

Are we missing anything major here? From our perspective it seems as if there is more stuff happening "underneath the surface" when manipulating the table via drag and drop but we can't quite put our finger on it. Do we need to save more information that needs to be restored to the table when loading? Or is there another workaround for this problem? MultiViewState only helps to keep changes in the Session, but we require persistence to the next session (login).

We are using PrimeFaces 11.0.0 / Extensions 11.0.1 on JBoss EAP 7.4.6 with JSF 2.3 (Mojarra 2.3.14) and sadly, we cannot upgrade PrimeFaces right now due to external conditions to test if this problem was already solved.

Many thanks in advance for reading.
Cheers

**Comment by melloware** [comment]
I wonder if there is a way you can plug into MVS: https://primefaces.github.io/primefaces/13_0_0/#/core/multiviewstate?id=multiviewstate  

It does all this already remember your sorts, filters, columns etc but right now just keeps it in the session storage.  I wonder if there is way you can plug in and persistent and restore from a Database.

@tandraschko @Rapster any thoughts?

**Comment by Rapster** [comment, marked as answer]
A very similar issue has been fixed recently: https://github.com/primefaces/primefaces/issues/10334
